### ISSUE TITLE
Add star-shaped geometries in 2D

### ISSUE THREAD

**Issue by pescap** [lululxvi/deepxde#1280]
It would be interesting to add a general star-shaped geometry. For star-shaped domains, one can set the radius as:
```
r(\theta) = r_0 + \sum_{j=1}^J [a_j\sin(j \theta)  + b_j \cos(j \theta)]
```
The input parameters could be: `(r_0, J, a_j, b_j)`.

I'll try to cook up something. 

https://github.com/lululxvi/deepxde/discussions/1264

**Comment by pescap** [comment]
Two options: 
- Take the circle, and apply the transformation to the point generator. It is quite common to do this, but the point distribution could result not perfectly homogeneous.
- Create a brand new class.

**Comment by pescap** [comment]
Required:
- `def on_boundary(self, x)`. Easy to define with the parametrization at hand
- `def inside`: Also straightforward to implement
-  `random_points`. Check which is the best manner to implement this method
-  `random_boundary_points`: Easy, with the parametrization.
```

Are we ok with these methods? Or do we need others?

**Comment by lululxvi** [comment]
Yes, it sounds good. These are basic APIs. `boundary_normal` is only needed for certain BCs.

**Comment by AudreyDemafo** [comment]
Hi @lululxvi and @pescap , I am facing a problem regarding the use of ``` boundary_normal ``` in my code. Here is the code 

```
Ta = 323.15 # Drying temperature
Xa = 0.40
R = 0.2
Re = 2300
X_ref = 0.388

def pde(x, y):
    T = y[:,0:1]
    X = y[:,1:2]
    P = y[:,2:3]
    dTdt = dde.grad.jacobian(y, x, i=0, j=0)
    dTdx = dde.grad.jacobian(y, x, i=0, j=1)
    dTdy = dde.grad.jacobian(y, x, i=0, j=2)
    rho_a = (341.25/(T + 273))
    c_a = 0.9774 + 0.1124e-3*(Ta + 273) + 1.19035e-7*(Ta + 273)**2
    c_air = 4e-5*(T + 273) + 0.0246
    mu_a = 1.66e-5*((T+273)/273)**0.756
    D_w = (T/273)**(1.81)
    grad_gradTx = dde.grad.jacobian(c_air*dTdx, x)
    grad_gradTy = dde.grad.jacobian(c_air*dTdy, x)
    
    dXdt = dde.grad.jacobian(y, x, i=1, j=0)
    global  a, b, c 
    a, b, c = 1.58772058e-11, 2.59623116e-14, 2.4943133761118243e-09
    a1 = 6.28509758e5
    b1 = 1.78292325e3
    c1 = 1.68500000e0
    d1 = 73824167.34861554
    a11 = 0.11960588
    b11 = 0.00017035
    c11 = 21.133862174623196
    D0 = -a1*Ta**3 + b1*Ta**2 - c1*Ta + d1
    z = -a11*Ta**2 + b11*Ta + c
    Deff = -a*Ta**2 + b*Ta + c
    ddXX = dde.grad.hessian(y, x, i=1, j=1, component=1)
    ddYY = dde.grad.hessian(y, x, i=1, j=2, component=1)
    diffusion = Deff*(ddXX + ddYY)
    T1= rho_a*c_a*dTdt - grad_gradTx - grad_gradTy
    M= dXdt - diffusion
    
    dPdt = dde.grad.jacobian(y, x, i=2, j=0)
    P= dPdt + ((D0*pow(10, (X_ref - X)/z))*P)/2.303
    
    return [T1, M, P]
   
    # the domain
geom = dde.geometry.Disk(center = (0,0), radius=0.5)
time = dde.geometry.TimeDomain(0, 60)
geomtime = dde.geometry.GeometryXTime(geom, time)

# Boundary
def boundary(x, on_boundary):
    return on_boundary and geom.on_boundary(x[:1])
def func_init(x, on_initial):
    return on_initial

def Neumann_T(x, T):
    (**I need help here**)
    #T = y[:, 0:1]
    c_air = 4e-5*(T + 273.15) + 0.0246
    c_a = 0.9774 + 0.1124e-3*(Ta + 273.15) + 1.19035e-7*(Ta + 273.15)**2
    m_a = 1.66e-5*((T+273.15)/273.15)**0.756
    k = ( 2 + 0.6*np.sqrt(Re)*(c_a*m_a/c_air)**(1/3) )*c_air/R 
    return c_air*geom.boundary_normal(x) - np.pi*0.5**2*k*(Ta - T)

def Neumann_X(x, X, T):
  (**I need help here**)
    # T = y[:, 0:1]
    # X = y[:, 1:2]
    Deff = a*Ta**2 + b*Ta + c
    D_w = (T/273.15)**(1.81)
    h = ( 2 + 0.6*np.sqrt(Re)*7**(1/3) )*D_w/R
    return -Deff*geom.boundary_normal(x) - np.pi*0.5**2*h*(Xa - X)

def func_init_T(x):
    return 288.15

def func_init_X(x):
    return 0.50

def func_init_P(x):
    return 0.45

bc1 = dde.NeumannBC(geomtime, Neumann_T, boundary, component=0)
bc2 = dde.NeumannBC(geomtime, Neumann_X, boundary, component=1)
bc3 = dde.IC(geomtime, func_init_T, func_init, component=0)
bc4 = dde.IC(geomtime, func_init_X, func_init, component=1)
bc5 = dde.IC(geomtime, func_init_P, func_init, component=2)

data = dde.data.TimePDE(geomtime,
                        pde,
                        [bc1, bc2, bc3, bc4, bc5],
                        num_domain = 50000,
                        num_boundary = 20000,
                        num_initial=20000,
                        num_test=30000
                    )

nnet = dde.maps.FNN([3] + [100]*30 + [3], "tanh", "Glorot uniform", layer_normalization="before")

model = dde.Model(data, nnet)

# early_stopping = dde.callbacks.EarlyStopping(min_delta=1e-3, patience=100) # patience: number of epochs with no improvement after which training will be stopped
# min_delta: If the loss decreases by at least 1e-3, it is considered as the model improvement.
loss_weights = [0.1, 0.1, 0.1, 1e-5, 0.1, 1e-5, 0.1, 0.1]

model.compile("adam", lr = 1e-3, loss="mse", loss_weights=loss_weights) 
 
# model.train(iterations=2000, display_every=100, callbacks=[early_stopping])
# model.compile("L-BFGS")

# losshistory, train_state = model.train(iterations=5000)

# dde.saveplot(
    # losshistory, train_state, issave=True, isplot=True)

# plt.savefig('Loss Function.png')

```

**The Neumann boundary conditions are respectively defined as:**
k*dT/dn = coefficient1*(Ta - T) // which T being one of the unknown
D_eff*dX/dn = coeffcicient2*(Xa - X) 

Please, kindly suggest. Thank you

**Comment by lululxvi** [comment]
You need to first figure out which line of code makes the error.

**Comment by pescap** [comment]
I initially used random point sampling based on a Disk, but the resulting sampling was not uniform. Therefore, I am now attempting to employ bbox-type random sampling to improve it.

Original attempt:

```
import numpy as np
from scipy import spatial
import deepxde 
from deepxde.geometry import Geometry
from deepxde.geometry.geometry_nd import Hypercube
from deepxde.geometry.sampler import sample
from deepxde import config
from deepxde.utils import isclose, vectorize


def polar(x):
    r = np.sqrt(x[:, 0] ** 2 + x[:, 0] ** 2)
    theta = np.arctan2(x[:, 1], x[:, 0])
    return r, theta

def pivot(x):
    return np.array([x[:, 1], -x[:, 0]]).T

class StarShaped(Geometry):
    def __init__(self, center, radius, coeffs_cos, coeffs_sin):
        self.center = np.array(center, dtype=config.real(np))
        self.radius = radius
        self.coeffs_cos = coeffs_cos
        self.coeffs_sin = coeffs_sin
        self.max_radius = radius + np.sum(coeffs_cos) + np.sum(coeffs_sin)
        super().__init__(2, (self.center - self.max_radius, self.center + self.max_radius), 2 * self.max_radius)

        self._r2 = radius**2

    def r_theta(self, theta):
        result = self.radius * np.ones(theta.shape)
        for i, (coeff_cos, coeff_sin) in enumerate(zip(self.coeffs_cos, self.coeffs_sin), start=1):
            result += coeff_cos * np.cos(i * theta) + coeff_sin * np.sin(i * theta)
        return result
    
    def dr_theta(self, theta):
        result = np.zeros(theta.shape)
        for i, (coeff_cos, coeff_sin) in enumerate(zip(self.coeffs_cos, self.coeffs_sin), start=1):
            result += - coeff_cos * i * np.sin(i * theta) + coeff_sin * i * np.cos(i * theta)
        return result
        
    def inside(self, x):
        r, theta = polar(x)
        r_theta = self.r_theta(theta)
        return r_theta <= r
        
    def on_boundary(self, x):
        r, theta = polar(x)
        r_theta = self.r_theta(theta)
        return isclose(np.linalg.norm(r_theta - r))

    def polar_vectors(self, x):
        e_r = x
        l = np.linalg.norm(e_r, axis=-1, keepdims=True)
        e_r = e_r / l
        e_theta = np.array([-e_r[:, 1], e_r[:, 0]]).T
        return e_r, e_theta
    
    def boundary_normal(self, x):
        _, theta = polar(x)
        dr_theta = self.dr_theta(theta)        
        r_theta = self.r_theta(theta)        
        
        dxt = np.vstack((dr_theta * np.cos(theta) - r_theta * np.sin(theta), dr_theta * np.sin(theta)+ r_theta * np.cos(theta))).T                
        l = np.linalg.norm(dxt, axis=-1, keepdims=True)
        dxt /= l
        dxn = pivot(dxt)
        return dxn
        
    def random_points(self, n, random="pseudo"):
        # http://mathworld.wolfram.com/DiskPointPicking.html
        rng = sample(n, 2, random)
        r, theta = rng[:, 0], 2 * np.pi * rng[:, 1]
        r_theta = self.r_theta(theta)
        x, y = np.cos(theta), np.sin(theta)
        return (np.sqrt(r) * r_theta * np.vstack((x, y))).T + self.center
        
    def uniform_boundary_points(self, n):
        theta = np.linspace(0, 2 * np.pi, num=n, endpoint=False)
        r_theta = self.r_theta(theta)
        X = np.vstack((r_theta * np.cos(theta), r_theta * np.sin(theta))).T
        return X + self.center

    def random_boundary_points(self, n, random="pseudo"):
        u = sample(n, 1, random)        
        theta = 2 * np.pi * u
        r_theta = self.r_theta(theta)
        X = np.hstack((r_theta * np.cos(theta), r_theta * np.sin(theta)))
        return X + self.center

```

**Comment by pescap** [comment]
For those that it might interest, this original setting yields the following result for

```
from matplotlib import pyplot as plt

center = [0, 0]
radius = 2
coeffs_sin = [.5, 0.25, 0.12, 0.08, 0.04]
coeffs_cos = [.5,.25,.12, .08, 0.04]

geom = StarShaped(center, radius, coeffs_cos, coeffs_sin) 
x = geom.random_points(5000)
xb = geom.uniform_boundary_points(100)
xn = geom.boundary_normal(xb)
xt = -pivot(xn)


fig, ax = plt.subplots(1,1,figsize=(8,8))

ax.scatter(x[:, 0], x[:, 1])
ax.scatter(xb[:, 0], xb[:, 1])

ax.quiver(xb[:, 0], xb[:, 1], xn[:, 0], xn[:, 1], angles ="uv", color="r")
#ax.quiver(xb[:, 0], xb[:, 1], xt[:, 0], xt[:, 1], angles ="uv", color="r")

plt.show(block=False)
```

The output:
![output](https://github.com/lululxvi/deepxde/assets/14826266/2c0f86ca-c833-4f8f-8234-8603fcb88b9e)

**Comment by pescap** [comment]
Works like a charm with bbox! Prepaing the PR.

![output (2)](https://github.com/lululxvi/deepxde/assets/14826266/54e0f063-496f-4107-aa5a-26f5eeaa69ba)
